PL2/PL2-shared/src/main/java/nl/tudelft/pl2016gr2/collectioninterfaces/GenomeIterator.java

```diff
@@ -0,0 +1,105 @@
+package nl.tudelft.pl2016gr2.collectioninterfaces;
+
+import nl.tudelft.pl2016gr2.model.Node;
+import nl.tudelft.pl2016gr2.model.OriginalGraph;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.NoSuchElementException;
+import java.util.Queue;
+
+/**
+ * This will provide an Iterator that can iterate over a single genome's path through the sequence
+ * graph.
+ * <p>
+ * Because finding the next element is more complex than in a generic <code>Collection</code>, this
+ * class is optimized to only search for the the next element once, in the {@link #hasNext()}
+ * method.
+ * Unlike the {@link Iterator} interface, it does not find new elements every time {@link #next()}
+ * is called.
+ * </p>
+ *
+ * @author Wouter Smit
+ */
+public class GenomeIterator implements Iterator<Node> {
+
+  private final OriginalGraph graph;
+  private final String genome;
+  private Node current;
+
+  /**
+   * Creates an iterator over the <code>graph</code> for <code>genome</code>.
+   * <p>
+   * Note that the genome must be present in the graph, otherwise a {@link
+   * NoSuchElementException} will be thrown.
+   * </p>
+   *
+   * @param graph  The graph to traverse through
+   * @param genome The genome to find and iterate over
+   */
+  public GenomeIterator(OriginalGraph graph, String genome) {
+    this.genome = genome;
+    this.graph = graph;
+    this.current = findFirstGenomeOccurrence();
```

Comment:
It's a little confusing to the user of this class to get the root in the constructor. You can, for example, instead ad an if statement to hasNext() to check whether current == null. If this is the case, you find the root, otherwise you just run the usual code.